src/basic/process-util.c

```diff
@@ -698,6 +698,62 @@ int wait_for_terminate_and_warn(const char *name, pid_t pid, bool check_exit_cod
         return -EPROTO;
 }
 
+int wait_for_terminate_with_timeout(pid_t pid, usec_t timeout) {
+        sigset_t mask;
+        int r;
+        siginfo_t status;
+        usec_t until;
+
+        assert_se(sigemptyset(&mask) == 0);
+        assert_se(sigaddset(&mask, SIGCHLD) == 0);
+
+        /* Drop into a sigtimewait-based timeout. Waiting for the
+         * remount pid to exit. */
+        until = now(CLOCK_MONOTONIC) + timeout;
+        for (;;) {
+                usec_t n;
+                struct timespec ts;
+
+                n = now(CLOCK_MONOTONIC);
+                if (n >= until)
+                        break;
+
+                zero(status);
+
+                r = sigtimedwait(&mask, NULL, timespec_store(&ts, until - n));
+                if (r < 0) {
+                        if (errno == EAGAIN) {
+                                /* Timed out, remount child is likely hung,
+                                 * but we can try. */
+                                if (waitid(P_PID, pid, &status, WEXITED|WNOHANG) == 0) {
+                                        if (status.si_pid == pid)
+                                                /* This is the correct child */
```

Comment:
This comment confused me.  In the case _below_, the same comment means "was the SIGCHILD for the correct child".  But I hadn't read that yet.  In this case, there was no SIGCHILD.  We are waiting for a specific PID, so we expect a result either for the correct child PID, or for no PID.

The easiest way to reduce confusion would be if we can just not do `waitid()` when we time out without getting a SIGCHILD.

Reply:
Sure, I can simplify and just let it fall through. Originally I was leaving a window open for a child process to exit in the timeout window, but no problem with closing that window that I can see.